From e519985b5dc5b3be51f1cf12eb37f5845b18b6bf Mon Sep 17 00:00:00 2001
From: Kito Cheng <kito.cheng@sifive.com>
Date: Fri, 23 Jan 2026 16:45:10 +0800
Subject: [PATCH] RISC-V: Add Zibi extension support

Add support for the Zibi (Branch on Immediate) extension which provides
beqi/bnei instructions for comparing a register with small immediates
(-1 or 1-31) and branching based on the result.

Changes:
- Add zibi extension definition to riscv-ext.def
- Add zibi constraint for matching valid immediates
- Add branch_on_immediate_operand predicate
- Add relational_comparison_operator predicate
- Split branch pattern into eqne (with zibi support) and relational variants
- Add arch-zibi.c and zibi.c test files
- Regenerate riscv-ext.opt and riscv-ext.texi

gcc/ChangeLog:

	* config/riscv/constraints.md (zibi): New constraint.
	* config/riscv/predicates.md (branch_on_immediate_operand): New predicate.
	(relational_comparison_operator): New predicate.
	* config/riscv/riscv-ext.def (zibi): New extension.
	* config/riscv/riscv-ext.opt: Regenerate.
	* config/riscv/riscv.md (ext): Add zibi.
	(ext_enabled): Add TARGET_ZIBI check.
	(*branch<mode>_eqne): New pattern with zibi support.
	(*branch<mode>_relational_comparison): New pattern for relational comparisons.
	* doc/riscv-ext.texi: Regenerate.

gcc/testsuite/ChangeLog:

	* gcc.target/riscv/arch-zibi.c: New test.
	* gcc.target/riscv/zibi.c: New test.
---
 gcc/config/riscv/constraints.md            |  5 ++
 gcc/config/riscv/predicates.md             |  9 +++
 gcc/config/riscv/riscv-ext.def             | 13 ++++
 gcc/config/riscv/riscv-ext.opt             |  2 +
 gcc/config/riscv/riscv.md                  | 40 ++++++++++++-
 gcc/doc/riscv-ext.texi                     |  4 ++
 gcc/testsuite/gcc.target/riscv/arch-zibi.c |  6 ++
 gcc/testsuite/gcc.target/riscv/zibi.c      | 70 ++++++++++++++++++++++
 8 files changed, 146 insertions(+), 3 deletions(-)
 create mode 100644 gcc/testsuite/gcc.target/riscv/arch-zibi.c
 create mode 100644 gcc/testsuite/gcc.target/riscv/zibi.c

diff --git a/gcc/config/riscv/constraints.md b/gcc/config/riscv/constraints.md
index 673b05954e1..4d063ece175 100644
--- a/gcc/config/riscv/constraints.md
+++ b/gcc/config/riscv/constraints.md
@@ -173,6 +173,11 @@
 (define_register_constraint "zmvr" "(TARGET_ZFA || TARGET_XTHEADFMV) ? GR_REGS : NO_REGS"
   "An integer register for  ZFA or XTheadFmv.")
 
+(define_constraint "zibi"
+  "An immediate that fits unsigned 5 bit or -1 for Zibi"
+  (and (match_code "const_int")
+       (match_test "TARGET_ZIBI && (IN_RANGE (ival, 1, 31) || (ival == -1))")))
+
 ;; Vector constraints.
 
 (define_register_constraint "vr" "TARGET_VECTOR ? V_REGS : NO_REGS"
diff --git a/gcc/config/riscv/predicates.md b/gcc/config/riscv/predicates.md
index 220a6f0830c..bad4f6500d2 100644
--- a/gcc/config/riscv/predicates.md
+++ b/gcc/config/riscv/predicates.md
@@ -236,6 +236,12 @@
   (and (match_code "const_int")
        (match_test "INTVAL (op) >= IMM_BITS - 1")))
 
+(define_predicate "branch_on_immediate_operand"
+  (ior (match_operand 0 "register_operand")
+       (ior (match_operand 0 "const_0_operand")
+	    (and (match_code "const_int")
+		 (match_test "TARGET_ZIBI && ((INTVAL (op) == -1) || IN_RANGE (INTVAL (op), 1, 31))")))))
+
 ;; A legitimate CONST_INT operand that takes more than one instruction
 ;; to load.
 (define_predicate "splittable_const_int_operand"
@@ -389,6 +395,9 @@
 (define_predicate "equality_operator"
   (match_code "eq,ne"))
 
+(define_predicate "relational_comparison_operator"
+  (match_code "lt,le,gt,ge,ltu,leu,gtu,geu"))
+
 (define_predicate "signed_order_operator"
   (match_code "eq,ne,lt,le,ge,gt"))
 
diff --git a/gcc/config/riscv/riscv-ext.def b/gcc/config/riscv/riscv-ext.def
index 1621e81b011..1ef1930d980 100644
--- a/gcc/config/riscv/riscv-ext.def
+++ b/gcc/config/riscv/riscv-ext.def
@@ -467,6 +467,19 @@ DEFINE_RISCV_EXT(
   /* BITMASK_BIT_POSITION*/ 8,
   /* EXTRA_EXTENSION_FLAGS */ 0)
 
+DEFINE_RISCV_EXT(
+  /* NAME */ zibi,
+  /* UPPERCASE_NAME */ ZIBI,
+  /* FULL_NAME */ "Branch on small immediate extension",
+  /* DESC */ "",
+  /* URL */ ,
+  /* DEP_EXTS */ ({}),
+  /* SUPPORTED_VERSIONS */ ({{0, 1}}),
+  /* FLAG_GROUP */ zi,
+  /* BITMASK_GROUP_ID */ BITMASK_NOT_YET_ALLOCATED,
+  /* BITMASK_BIT_POSITION*/ BITMASK_NOT_YET_ALLOCATED,
+  /* EXTRA_EXTENSION_FLAGS */ 0)
+
 DEFINE_RISCV_EXT(
   /* NAME */ zmmul,
   /* UPPERCASE_NAME */ ZMMUL,
diff --git a/gcc/config/riscv/riscv-ext.opt b/gcc/config/riscv/riscv-ext.opt
index 18402ea97c7..7b3317360d2 100644
--- a/gcc/config/riscv/riscv-ext.opt
+++ b/gcc/config/riscv/riscv-ext.opt
@@ -164,6 +164,8 @@ Mask(ZIMOP) Var(riscv_zi_subext)
 
 Mask(ZILSD) Var(riscv_zi_subext)
 
+Mask(ZIBI) Var(riscv_zi_subext)
+
 Mask(ZMMUL) Var(riscv_zm_subext)
 
 Mask(ZA128RS) Var(riscv_za_subext)
diff --git a/gcc/config/riscv/riscv.md b/gcc/config/riscv/riscv.md
index 3fe0ad0ccdf..7ea6ee704c3 100644
--- a/gcc/config/riscv/riscv.md
+++ b/gcc/config/riscv/riscv.md
@@ -291,7 +291,7 @@
 	(const_string "no")))
 
 ;; ISA attributes.
-(define_attr "ext" "base,f,d,vector"
+(define_attr "ext" "base,f,d,vector,zibi"
   (const_string "base"))
 
 ;; True if the extension is enabled.
@@ -310,6 +310,10 @@
 	 (and (eq_attr "ext" "vector")
 	      (match_test "TARGET_VECTOR"))
 	 (const_string "yes")
+
+	 (and (eq_attr "ext" "zibi")
+	      (match_test "TARGET_ZIBI"))
+	 (const_string "yes")
 	]
 	(const_string "no")))
 
@@ -3280,10 +3284,40 @@
 }
 [(set_attr "type" "branch")])
 
-(define_insn "*branch<mode>"
+(define_insn "*branch<mode>_eqne"
   [(set (pc)
 	(if_then_else
-	 (match_operator 1 "ordered_comparison_operator"
+	 (match_operator 1 "equality_operator"
+			 [(match_operand:X 2 "register_operand"            " r,   r")
+			  (match_operand:X 3 "branch_on_immediate_operand" "rJ,zibi")])
+	 (label_ref (match_operand 0 "" ""))
+	 (pc)))]
+  "!TARGET_XCVBI"
+{
+  bool long_jump_p = get_attr_length (insn) == 12;
+  if (which_alternative == 1)
+    {
+      if (long_jump_p)
+	return "b%N1i\t%2,%z3,1f; jump\t%l0,ra; 1:";
+
+      return "b%C1i\t%2,%z3,%l0";
+    }
+  else
+    {
+      if (long_jump_p)
+	return "b%N1\t%2,%z3,1f; jump\t%l0,ra; 1:";
+
+      return "b%C1\t%2,%z3,%l0";
+    }
+}
+  [(set_attr "type" "branch")
+   (set_attr "mode" "none")
+   (set_attr "ext" "base,zibi")])
+
+(define_insn "*branch<mode>_relational_comparison"
+  [(set (pc)
+	(if_then_else
+	 (match_operator 1 "relational_comparison_operator"
 			 [(match_operand:X 2 "register_operand" "r")
 			  (match_operand:X 3 "reg_or_0_operand" "rJ")])
 	 (label_ref (match_operand 0 "" ""))
diff --git a/gcc/doc/riscv-ext.texi b/gcc/doc/riscv-ext.texi
index 36b4dff31ff..bf997bad02f 100644
--- a/gcc/doc/riscv-ext.texi
+++ b/gcc/doc/riscv-ext.texi
@@ -134,6 +134,10 @@
 @tab 1.0
 @tab Load/store pair instructions extension
 
+@item @samp{zibi}
+@tab 0.1
+@tab Branch on small immediate extension
+
 @item @samp{zmmul}
 @tab 1.0
 @tab Integer multiplication extension
diff --git a/gcc/testsuite/gcc.target/riscv/arch-zibi.c b/gcc/testsuite/gcc.target/riscv/arch-zibi.c
new file mode 100644
index 00000000000..1d71f738611
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/arch-zibi.c
@@ -0,0 +1,6 @@
+/* { dg-do compile } */
+/* { dg-options "-march=rv64i_zibi0p1 -mabi=lp64" } */
+int foo()
+{
+}
+
diff --git a/gcc/testsuite/gcc.target/riscv/zibi.c b/gcc/testsuite/gcc.target/riscv/zibi.c
new file mode 100644
index 00000000000..aa0ce1b0a5e
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/zibi.c
@@ -0,0 +1,70 @@
+/* { dg-do compile } */
+/* { dg-skip-if "" { *-*-* } { "-O0" "-flto" } { "" } } */
+/* { dg-options "-march=rv64gc_zibi -mabi=lp64 -mtune=rocket" } */
+/* { dg-final { check-function-bodies "**" "" } } */
+
+int foo_n1(int a, int b, int c)
+{
+/*
+** foo_n1:
+**   ...
+**   b(eq|ne)i	[ats][0-9]+,-1,.L[0-9]+
+**   ...
+*/
+    if (a == -1)
+      return b - c;
+    return c - b;
+}
+
+int foo_zero(int a, int b, int c)
+{
+/*
+** foo_zero:
+**   ...
+**   b(eq|ne)	[ats][0-9]+,zero,.L[0-9]+
+**   ...
+*/
+    if (a == 0)
+      return b - c;
+    return c - b;
+}
+
+int foo_1(int a, int b, int c)
+{
+/*
+** foo_1:
+**   ...
+**   b(eq|ne)i	[ats][0-9]+,1,.L[0-9]+
+**   ...
+*/
+    if (a == 1)
+      return b - c;
+    return c - b;
+}
+
+int foo_31(int a, int b, int c)
+{
+/*
+** foo_31:
+**   ...
+**   b(eq|ne)i	[ats][0-9]+,31,.L[0-9]+
+**   ...
+*/
+    if (a == 31)
+      return b - c;
+    return c - b;
+}
+
+int foo_32(int a, int b, int c)
+{
+/*
+** foo_32:
+**   ...
+**   b(eq|ne)	[ats][0-9]+,[ats][0-9]+,.L[0-9]+
+**   ...
+*/
+    if (a == 32)
+      return b - c;
+    return c - b;
+}
+
-- 
2.47.3

